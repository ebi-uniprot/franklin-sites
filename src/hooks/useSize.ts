// NOTE: same as in uniprot-website codebase
import { MutableRefObject, useCallback, useEffect, useState } from 'react';

/**
 * given a reference to an actual HTML element, gives us access to its size
 * even through resizes and zoom
 * @param {MutableRefObject<HTMLElement>} ref
 * @returns {Array} tuple - return tuple
 * @returns {DOMRect | null} tuple[0]
 *          DOMRect object with the sizes of the passed ref, or null if no ref
 * @returns {() => void} tuple[1]
 *          Manual refresh function, to call if the size needs to be recomputed
 *          for whatever reason (e.g. content changed)
 */
function useSize(
  ref: MutableRefObject<HTMLElement | null>
): [rect: DOMRect | undefined, onResize: () => void] {
  const [rect, setRect] = useState<DOMRect>();

  const onResize = useCallback(() => {
    if (!ref.current) {
      return;
    }
    setRect(ref.current.getBoundingClientRect());
  }, [ref, setRect]);

  // This effect will run after EVERY render
  useEffect(() => {
    // ask for size value until we finally get one
    if (!rect) {
      onResize();
    }
  }); // No dependency array, it's on purpose!

  useEffect(() => {
    onResize(); // first time
    window.addEventListener('resize', onResize);
    // eslint-disable-next-line consistent-return
    return () => window.removeEventListener('resize', onResize);
  }, [ref, onResize]);

  return [rect, onResize];
}

export default useSize;
