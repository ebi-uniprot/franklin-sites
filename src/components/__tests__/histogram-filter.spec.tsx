import { render, fireEvent, screen } from '@testing-library/react';

import { HistogramFilter } from '..';

import useSize from '../../hooks/useSize';

jest.mock('../../hooks/useSize', () => jest.fn());
(useSize as jest.Mock).mockImplementation(() => [{ width: 250, height: 100 }]);

describe('HistogramFilter component', () => {
  const onChange = jest.fn();

  let rendered: ReturnType<typeof render>;

  const values = [...Array(10).keys()];

  beforeEach(() => {
    rendered = render(
      <HistogramFilter
        onChange={onChange}
        values={values}
        nBins={5}
        selectedRange={[0, 9]}
      />
    );
  });

  it('should render', () => {
    const { asFragment } = rendered;
    expect(asFragment()).toMatchSnapshot();
  });

  it('should call onChange when text input is provided', async () => {
    const startInput = screen.getByDisplayValue('0');
    const endInput = screen.getByDisplayValue('9');
    // narrow the range from the start
    fireEvent.change(startInput, { target: { value: '1' } });
    expect(onChange).toHaveBeenLastCalledWith([1, 9]);
    rendered.rerender(
      <HistogramFilter
        onChange={onChange}
        values={values}
        nBins={5}
        selectedRange={[1, 9]}
      />
    );
    // narrow the range from the start
    fireEvent.change(endInput, { target: { value: '5' } });
    expect(onChange).toHaveBeenLastCalledWith([1, 5]);
    rendered.rerender(
      <HistogramFilter
        onChange={onChange}
        values={values}
        nBins={5}
        selectedRange={[1, 5]}
      />
    );
    // invalid value
    onChange.mockClear(); // reset on change as we don't expect it to be called
    fireEvent.change(startInput, { target: { value: '-500' } });
    fireEvent.change(endInput, { target: { value: 'abc' } });
    // display invalid value
    expect(startInput).toHaveDisplayValue('-500');
    expect(endInput).toHaveDisplayValue('abc');
    // but revert to previous valid value on blur
    fireEvent.blur(startInput);
    fireEvent.blur(endInput);
    expect(startInput).toHaveDisplayValue('1');
    expect(endInput).toHaveDisplayValue('5');
    expect(onChange).not.toHaveBeenCalled();
  });
});
