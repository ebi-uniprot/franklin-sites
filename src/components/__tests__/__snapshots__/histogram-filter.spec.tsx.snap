// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HistogramFilter component should render 1`] = `
<DocumentFragment>
  <div
    class="histogram-filter"
  >
    <div
      class="histogram"
      data-testid="histogram"
    >
      <div
        class="histogram__bar-container"
        style="height: 50px;"
      >
        <div
          class="histogram histogram__bar histogram__bar--within-range"
          data-testid="histogram-bar"
          style="transform: scaleY(0.4); transition-delay: 0ms;"
        />
        <div
          class="histogram histogram__bar histogram__bar--within-range"
          data-testid="histogram-bar"
          style="transform: scaleY(0.4); transition-delay: 100ms;"
        />
        <div
          class="histogram histogram__bar histogram__bar--within-range"
          data-testid="histogram-bar"
          style="transform: scaleY(0.4); transition-delay: 200ms;"
        />
        <div
          class="histogram histogram__bar histogram__bar--within-range"
          data-testid="histogram-bar"
          style="transform: scaleY(0.4); transition-delay: 300ms;"
        />
        <div
          class="histogram histogram__bar histogram__bar--within-range"
          data-testid="histogram-bar"
          style="transform: scaleY(0.4); transition-delay: 400ms;"
        />
      </div>
      <svg
        height="100%"
        width="100%"
      >
        <g
          fill="none"
          pointer-events="all"
        >
          <rect
            class="overlay"
            cursor="crosshair"
            height="100"
            pointer-events="all"
            width="250"
            x="0"
            y="0"
          />
          <rect
            class="selection"
            cursor="move"
            fill="#777"
            fill-opacity="0.3"
            shape-rendering="crispEdges"
            stroke="#fff"
            style="display: none;"
          />
          <rect
            class="handle handle--w"
            cursor="ew-resize"
            style="display: none;"
          />
          <rect
            class="handle handle--e"
            cursor="ew-resize"
            style="display: none;"
          />
        </g>
      </svg>
    </div>
    <div
      class="histogram-filter__text-input-container"
    >
      <input
        style="width: 3ch;"
        type="text"
        value="0"
      />
      <input
        style="width: 3ch;"
        type="text"
        value="9"
      />
    </div>
  </div>
</DocumentFragment>
`;
